src/math/matrix.c

```diff
@@ -338,19 +338,32 @@ void Matrix_Interpolate(void)
     MATRIX *mptr = g_MatrixPtr;
     MATRIX *iptr = m_IMMatrixPtr;
 
-    if (m_IMRate == 2) {
-        mptr->_00 = (mptr->_00 + iptr->_00) / 2;
-        mptr->_01 = (mptr->_01 + iptr->_01) / 2;
-        mptr->_02 = (mptr->_02 + iptr->_02) / 2;
-        mptr->_03 = (mptr->_03 + iptr->_03) / 2;
-        mptr->_10 = (mptr->_10 + iptr->_10) / 2;
-        mptr->_11 = (mptr->_11 + iptr->_11) / 2;
-        mptr->_12 = (mptr->_12 + iptr->_12) / 2;
-        mptr->_13 = (mptr->_13 + iptr->_13) / 2;
-        mptr->_20 = (mptr->_20 + iptr->_20) / 2;
-        mptr->_21 = (mptr->_21 + iptr->_21) / 2;
-        mptr->_22 = (mptr->_22 + iptr->_22) / 2;
-        mptr->_23 = (mptr->_23 + iptr->_23) / 2;
+    if (m_IMRate == 2 || (m_IMFrac == 2 && m_IMRate == 4)) {
+        mptr->_00 += (iptr->_00 - mptr->_00) >> 1;
+        mptr->_01 += (iptr->_01 - mptr->_01) >> 1;
+        mptr->_02 += (iptr->_02 - mptr->_02) >> 1;
+        mptr->_03 += (iptr->_03 - mptr->_03) >> 1;
+        mptr->_10 += (iptr->_10 - mptr->_10) >> 1;
+        mptr->_11 += (iptr->_11 - mptr->_11) >> 1;
+        mptr->_12 += (iptr->_12 - mptr->_12) >> 1;
+        mptr->_13 += (iptr->_13 - mptr->_13) >> 1;
+        mptr->_20 += (iptr->_20 - mptr->_20) >> 1;
+        mptr->_21 += (iptr->_21 - mptr->_21) >> 1;
+        mptr->_22 += (iptr->_22 - mptr->_22) >> 1;
+        mptr->_23 += (iptr->_23 - mptr->_23) >> 1;
+    } else if (m_IMFrac == 1) {
+        mptr->_00 += (iptr->_00 - mptr->_00) >> 2;
+        mptr->_01 += (iptr->_01 - mptr->_01) >> 2;
+        mptr->_02 += (iptr->_02 - mptr->_02) >> 2;
+        mptr->_03 += (iptr->_03 - mptr->_03) >> 2;
+        mptr->_10 += (iptr->_10 - mptr->_10) >> 2;
+        mptr->_11 += (iptr->_11 - mptr->_11) >> 2;
+        mptr->_12 += (iptr->_12 - mptr->_12) >> 2;
+        mptr->_13 += (iptr->_13 - mptr->_13) >> 2;
+        mptr->_20 += (iptr->_20 - mptr->_20) >> 2;
+        mptr->_21 += (iptr->_21 - mptr->_21) >> 2;
+        mptr->_22 += (iptr->_22 - mptr->_22) >> 2;
+        mptr->_23 += (iptr->_23 - mptr->_23) >> 2;
```

Comment:
I don't quite get these changes, are they necessary? Because they seem like a hand-rolled optimization of (x*m_IMFrac/m_IMRate) operation.

Reply:
This is the result without these changes.

https://www.youtube.com/watch?v=9UaecY3xYPU

I believe it resolves an overflow issue (but I don't fully understand it I have to admit); I don't know if this affects anything in the current release. Perhaps we should make a separate function just for hair interpolation to play it safe?

Reply:
Nah it's good, since you tested it and that's what Tomb3 uses anyway.